crypto/dh/dh_key.c

```diff
@@ -230,3 +230,49 @@ static int dh_finish(DH *dh)
     BN_MONT_CTX_free(dh->method_mont_p);
     return 1;
 }
+
+int dh_buf2key(DH *dh, const unsigned char *buf, size_t len)
+{
+    BIGNUM *pubkey;
+
+    if ((pubkey = BN_bin2bn(buf, len, NULL)) == NULL
+        || BN_is_zero(pubkey)
+        || !DH_set0_key(dh, pubkey, NULL))
+        goto err;
+    return 1;
+err:
+    DHerr(DH_F_DH_BUF2KEY, ERR_R_INTERNAL_ERROR);
+    BN_free(pubkey);
+    return 0;
+}
+
+size_t dh_key2buf(const DH *dh, unsigned char **pbuf_out)
+{
+    unsigned char *pbuf;
+    const BIGNUM *p;
+    const BIGNUM *pubkey;
+    int pubkey_size;
+    int pad = 0;
+
+    DH_get0_pqg(dh, &p, NULL, NULL);
+    DH_get0_key(dh, &pubkey, NULL);
+    if (p == NULL || pubkey == NULL
+            || (pubkey_size = BN_num_bytes(pubkey)) == 0) {
```

Comment:
Not sure why we need to calculate `pubkey_size` and `pad` at all. Don't we just need to know the size of the buffer (`BN_num_bytes(p)`)?